src/Microsoft.NetCore.Analyzers/CSharp/Performance/CSharpMutableStructsShouldNotBeUsedForReadonlyFields.Fixer.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Composition;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.NetCore.Analyzers;
+
+namespace Microsoft.NetCore.CSharp.Analyzers.Performance
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp), Shared]
+    public sealed class CSharpMutableStructsShouldNotBeUsedForReadonlyFieldsFixer : NetCore.Analyzers.Performance.MutableStructsShouldNotBeUsedForReadonlyFieldsFixer
+    {
+        protected override void AnalyzeCodeFix(CodeFixContext context, SyntaxNode targetNode)
+        {
+            if (!(targetNode is FieldDeclarationSyntax fieldDeclarationSyntax))
+            {
+                return;
+            }
+
+            var readonlyModifiers =
+                fieldDeclarationSyntax.Modifiers.Where(modifier => modifier.IsKind(SyntaxKind.ReadOnlyKeyword)).ToArray();
+
+            if (!readonlyModifiers.Any())
+            {
+                return;
+            }
+
+            var removeReadonlyAction = CodeAction.Create(
+                MicrosoftNetCoreAnalyzersResources.MutableStructsShouldNotBeUsedForReadonlyFieldsTitle,
+                async ct => await RemoveReadonlyKeyword(context, fieldDeclarationSyntax).ConfigureAwait(false),
+                EquivalencyKey);
+
+            context.RegisterCodeFix(removeReadonlyAction, context.Diagnostics);
+        }
+
+        private static async Task<Document> RemoveReadonlyKeyword(CodeFixContext context,
+            FieldDeclarationSyntax fieldDeclarationSyntax)
+        {
+            var editor = await DocumentEditor.CreateAsync(context.Document, context.CancellationToken).ConfigureAwait(false);
+            var withoutReadonly = fieldDeclarationSyntax.WithModifiers(
```

Comment:
A slighly simpler implementation would be: http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Features/CodeFixes/HideBase/HideBaseCodeFixProvider.AddNewKeywordAction.cs,37

This also allows you to make the fixer language agnostic (you don't need language specific `SyntaxKind` anymore)

Reply:
I tried approaching it as you suggested but I think I'm missing something. The situation:

1. The analyzer now operates on symbols and reports the location of the diagnostic as `fieldSymbol.Locations.First()`. This means the location of the symbol name is reported.
2. That means the the fix will get a span that points to a `VariableDeclaratorSyntax`, not the actual `FieldDeclarationSyntax`. That works fine for the `SyntaxGenerator`, which ascends the tree correctly and returns a new `FieldDeclarationSyntax` node, but then I can't simply call `root.ReplaceNode` because the nodes are of different types.
3. Furthermore, this complicates executing the fix on statement `private readonly SpinLock _sl1 = new SpinLock(), sl2 = new SpinLock();`. Here the generator doesn't work because it creates a new `FieldDeclarationSyntax` but *without* the second variable declaration.

Either I'm missing the correct API/usage of it or it looks like it will be easier to make the fix language-specific and simply use:

``` csharp
root.FindNode(context.Span).FirstAncestorOrSelf<FieldDeclarationSyntax>() // this is language-specific
...
var generator = SyntaxGenerator.GetGenerator(context.Document);
var nodeWithoutReadonly = generator.WithModifiers(targetNode, generator.GetModifiers(targetNode).WithIsReadOnly(false));
var newRoot = root.ReplaceNode(originalFieldDeclaration, nodeWithoutReadonly);
return Task.FromResult(context.Document.WithSyntaxRoot(newRoot));
```

Testing this, it leads to correct behavior even in case of multi-declaration.

Reply:
Thanks, I meant to recommend the exact same implementation as your comment, but keep it in the shared layer with a type parameter TFieldDeclarationSyntax, where TFieldDeclarationSyntax: SyntaxNode. Language specific subtypes will just need to provide the language specific type argument FieldDeclarationSyntax and have no code within.